File: core/r/lib/clean.R
# Minimal, logged cleaning: only what is needed to run the analysis, with every
# step recorded so the raw-to-analyzed path stays auditable.

clean_tidy <- function(df, spec) {
  steps <- character(0)
  n0 <- nrow(df)
  x <- spec$data$x
  y <- spec$data$y
  if (!x %in% names(df)) stop(sprintf("column '%s' not found in data", x))
  if (!y %in% names(df)) stop(sprintf("column '%s' not found in data", y))

  yv <- suppressWarnings(as.numeric(df[[y]]))
  coerced <- sum(is.na(yv) & !is.na(df[[y]]))
  if (coerced > 0) {
    steps <- c(steps, sprintf("Coerced column '%s' to numeric; %d value(s) became NA.", y, coerced))
  }
  df[[y]] <- yv

  keep <- !is.na(df[[x]]) & is.finite(df[[y]])
  dropped <- sum(!keep)
  if (dropped > 0) {
    steps <- c(steps, sprintf("Dropped %d row(s) with missing '%s' or '%s' (%d -> %d rows).",
                              dropped, x, y, n0, sum(keep)))
  }
  df <- df[keep, , drop = FALSE]

  if (length(steps) == 0) steps <- "No cleaning applied; input used as-is."
  list(df = df, steps = steps)
}

File: core/python/pubready/clean.py
"""Minimal, logged cleaning for x/y recipes (mirrors the R clean_tidy)."""
import numpy as np
import pandas as pd


def clean_xy(df, x, y):
    df = df.copy()
    steps = []
    n0 = len(df)
    if x not in df.columns:
        raise ValueError(f"column '{x}' not found in data")
    if y not in df.columns:
        raise ValueError(f"column '{y}' not found in data")

    yv = pd.to_numeric(df[y], errors="coerce")
    coerced = int((yv.isna() & df[y].notna()).sum())
    if coerced:
        steps.append(f"Coerced column '{y}' to numeric; {coerced} value(s) became NA.")
    df[y] = yv

    mask = df[x].notna() & df[y].notna() & np.isfinite(df[y])
    dropped = int((~mask).sum())
    if dropped:
        steps.append(f"Dropped {dropped} row(s) with missing '{x}' or '{y}' "
                     f"({n0} -> {int(mask.sum())} rows).")
    df = df[mask].reset_index(drop=True)

    if not steps:
        steps = ["No cleaning applied; input used as-is."]
    return df, steps
